### PR TITLE
feat(auth): add reset password

### PR DESCRIPTION
**Pull request by sktimalsina**
- [x] PR title and description conform to [Pull Request](https://github.com/aws-amplify/amplify-android/blob/main/CONTRIBUTING.md#pull-request-guidelines) guidelines.

*Issue #, if available:*

*Description of changes:*
Add reset password use case.

*How did you test these changes?*
(Please add a line here how the changes were tested)
Tested e2e using a sample app and verifying that an email for reset password was sent out.

- [x] Added Unit Tests
- [ ] Tested on frontend (Add Screenshots)
- [ ] Successful logs (Attach them to the PR)

*Documentation update required?*
- [ x] No
- [ ] Yes (Please include a PR link for the documentation update)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by tylerjroach — COMMENTED** [review]
Tests look good. Thank you for focus on that!

**Review by gpanshu — DISMISSED** [review]

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/AWSCognitoAuthPlugin.kt [hunk 1]
```diff
@@ -717,21 +720,32 @@ class AWSCognitoAuthPlugin : AuthPlugin<AWSCognitoAuthServiceBehavior>() {
...
 
+    @OptIn(DelicateCoroutinesApi::class)
     override fun resetPassword(
         username: String,
         options: AuthResetPasswordOptions,
         onSuccess: Consumer<AuthResetPasswordResult>,
         onError: Consumer<AuthException>
     ) {
-        TODO("Not yet implemented")
+        val cognitoIdentityProviderClient = configureCognitoClients().cognitoIdentityProviderClient ?: return
```
Comment (div5yesh) [inline aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/AWSCognitoAuthPlugin.kt:730]: Failing silently is not a good option.

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/usecases/ResetPasswordUseCase.kt [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+                            mapOf(),
+                            response.codeDeliveryDetails.toAuthCodeDeliveryDetails()
+                        )
+                    )
+                )
+            }
+        } catch (ex: Exception) {
+            withContext(Dispatchers.Main) {
+                onError.accept(
+                    AuthException(
```
Comment (div5yesh) [inline aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/usecases/ResetPasswordUseCase.kt:65]: Please use `CognitoExceptionConvertor.lookup`.

File: aws-auth-cognito/src/test/java/com/amplifyframework/auth/cognito/usecases/ResetPasswordUseCaseTest.kt [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+            clientId = dummyClientId
+        }
+
+        // WHEN
+        runBlocking {
+            resetPasswordUseCase.execute(dummyUserName, AuthResetPasswordOptions.defaults(), {}, {})
+        }
+
+        // THEN
+        assertEquals(stubPasswordRequest, forgotPassWordRequest)
```
Comment (div5yesh) [inline aws-auth-cognito/src/test/java/com/amplifyframework/auth/cognito/usecases/ResetPasswordUseCaseTest.kt:95]: The objects are hardcoded, will they be always equal?

**Review by div5yesh — COMMENTED** [review]
Few comments.

**Review by tylerjroach — APPROVED** [review]

**Pull request merged**